lint/importShadow_checker.go

```diff
@@ -42,18 +42,27 @@ type importShadowChecker struct {
 }
 
 func (c *importShadowChecker) VisitLocalDef(def astwalk.Name, _ ast.Expr) {
-	for _, imp := range c.ctx.Context.pkg.Imports() {
-		if imp.Name() == def.ID.Name {
-			c.warnImportShadowed(def.ID, imp)
+	for _, v := range c.ctx.Context.typesInfo.Defs {
```

Comment:
I'm a bit concerned that it iterated all definitions for every local definition.
This can be noticeable when checking big projects like 
`gocritic check-project -enable=importShadow $GOROOT/src`
I think it will take measurable amount of time (we have `unusedParam` that works several minutes vs 1-2secs of other checkers).
I'll think about this after working hours, maybe there is a better 
We can still merge this solution, but it would be also good to either add TODO notice there or create an issue that this checker may be slow and could benefit from tuning.

Reply:
Completely agree, currently it is O(n^2) for count of definitions - and it is really bad.

Maybe we should add some "precomputation" stage, for example it could look like - each linter implements some function that is called for example for each package or for each file, during this stage we will calculate all supplemental maps/structures which then will be used in the real linter checkers.

Reply:
This is 3rd checker that can benefit from precomputed data.
The only issue is that it's not optimal to precompute all such data without knowing it's 100% needed, so we need to know what kind of resources checker needs, and if specific indexes has no usages (because related checkers are not enabled during the run), we should not load them.

For now, it can be simpler to track all such linters and make a solution that fits them all later.

I would also like to think about this particular check. We may be lacking some knowledge that could make O(n^2) not needed at all, even without precomputed map.

(If @cristaloleg approves, we may merge as is and improve on it later.)

Reply:
I have some proposal for this precomputations it looks like this:
1) We define some Option flags for the linters, like PrefetchImportedPackages, etc...
2) During linter registration it can define as many options as it needs, options could be file or package specific
3) Before real running linters in the goroutines, all precomputations needed are summarized and done which results in more filled context
4) linters use precomputed data from the context

It is also possible to "hack"-fix current approach, we can do this by doing not real caching, for example - in the beginning of the linter check we check local cache object if current file was already precomputed, and if so use this precomputed (in my case - map[file path]map[package name]bool structure)

From my perspective second approach is not very good cause precomputed data can't be shared between different linters, but for one-point-optimization it looks ok.